In [32]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori,association_rules


In [18]:
df = pd.read_csv(r"C:\Users\SURAJ\Downloads\Groceries.csv",header=None,names=['Products'])
df.head()

Products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [19]:
df.shape

(20, 1)

* applying lambda function we slpit the products by comma

In [21]:
data = list(df["Products"].apply(lambda x:x.split(',')))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

* ENCODING

In [26]:
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_data = te.fit_transform(data)
df_ = pd.DataFrame(te_data,columns=te.columns_).astype(int)
df_.head()

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  ...  JAM  MAGGI  MILK  SUGER  TEA
0        1          0      1     0       0  ...    0      0     1      0    0
1        1          0      1     0       0  ...    0      0     1      0    0
2        0          1      1     0       0  ...    0      0     0      0    1
3        0          0      1     0       0  ...    1      1     1      0    0
4        1          0      0     0       0  ...    0      1     0      0    1

[5 rows x 11 columns]

* Applying Apriori and Resulting

    We can change all the parameters in the Apriori Model in the mlxtend package.
    We will try to use minimum support parameters for this modeling.
    For this, I set a min_support value with a threshold value of 20% .


In [33]:
df_new=apriori(df_, min_support=0.2, use_colnames=True)
df_new

support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BREAD, BISCUIT)
10     0.20         (MILK, BREAD)
11     0.20        (BREAD, SUGER)
12     0.20          (BREAD, TEA)
13     0.20  (CORNFLAKES, COFFEE)
14     0.20       (SUGER, COFFEE)
15     0.20          (MAGGI, TEA)

In [34]:
df_ar = association_rules(df_new, metric = "confidence", min_threshold = 0.6)
df_ar

antecedents consequents  ...  conviction  zhangs_metric
0        (MILK)     (BREAD)  ...        1.75       0.250000
1       (SUGER)     (BREAD)  ...        1.05       0.035714
2  (CORNFLAKES)    (COFFEE)  ...        1.80       0.571429
3       (SUGER)    (COFFEE)  ...        1.80       0.571429
4       (MAGGI)       (TEA)  ...        3.25       0.750000

[5 rows x 10 columns]

For example, if we examine our 1st index value;

    * The probability of seeing sugar sales is seen as 30%.
    * Bread intake is seen as 65%.
    * We can say that the support of both of them is measured as 20%.
    * 67% of those who buys sugar, buys bread as well.
    * Users who buy sugar will likely consume 3% more bread than users who don't buy sugar.
    * Their correlation with each other is seen as 1.025.
